### DISCUSSION TITLE
How to handle Dynamic server usage in next build without i18n routing

### DISCUSSION THREAD

**Discussion by MHMDHIDR** [amannn/next-intl#1551] [Q&A] [answered]
I'm building a side project, when I try to build it I get the following error:

> Error getting user language: Error: Dynamic server usage: Route /dashboard/users couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
>     at g (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/3476.js:2:87443)
>     at u (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/3476.js:3:47101)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/3476.js:405:59645)
>     at i (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/6535.js:1:3490)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/6535.js:1:33225)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/7488.js:2:97665)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:124:3912)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/7488.js:2:98209)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:124:3912)
>     at <unknown> (/Applications/XAMPP/xamppfiles/htdocs/crm/.next/server/chunks/7488.js:2:98407) {
>   description: "Route /dashboard/users couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error",
>   digest: 'DYNAMIC_SERVER_USAGE'
> }

I'm using NextJS 15, and this is my routing.ts config:
```ts
export type Locale = (typeof locales)[number]

export const locales = ['en', 'ar'] as const
export const defaultLocale: Locale = 'en'

export const routing = defineRouting({
  locales,
  defaultLocale: locales[0],
  localePrefix: 'never'
})
```

I'm pretty sure I'm following as best as I can using the Docs, I need to get the user locale from a database value 'en' or 'ar'.

I've even used setRequestLocale in layout.tsx:

```ts
export default async function RootLayout({ children }: { children: React.ReactNode }) {
  const locale = (await getUserLanguage()) ?? (await getLocale())
  const messages = await getMessages({ locale })
  setRequestLocale(locale)
```

How do I solve the errors.
For more info this is my repo:-
https://github.com/MHMDHIDR/crm

**Comment by amannn** [comment, marked as answer]
> Error: Dynamic server usage: Route /dashboard/users couldn't be rendered statically because it used headers. See more info here: https://nextjs.org/docs/messages/dynamic-server-error

Your `getUserLanguage` function relies on reading headers, therefore a page can naturally not be static at the same time—I guess that makes sense. For apps behind a login dynamic rendering is quite common.

**Reply by MHMDHIDR** [reply]
So it quite OK to have this behavior or not statically rendered APPs if my app requires login to render the translated pages?

**Reply by amannn** [reply]
Yes!
